pep-0544.txt

```diff
@@ -0,0 +1,967 @@
+the subtype relationship is implicit and only via structural
+subtyping -- the semantics of inheritance is not changed. Examples::
+
+    class PColor(Protocol):
+        @abstractmethod
+        def draw(self) -> str:
+            ...
+        def complex_method(self) -> int:
+            # some complex code here
+
+    class NiceColor(PColor):
+        def draw(self) -> str:
+            return "deep blue"
+
+    class BadColor(PColor):
+        def draw(self) -> str:
+            return super().draw()  # Error, no default implementation
+
+    class ImplicitColor:   # Note no 'PColor' base here
+        def draw(self) -> str:
+            return "probably gray"
+        def comlex_method(self) -> int:
+            # class needs to implement this
+
+    nice: NiceColor
+    another: ImplicitColor
+
+    def represent(c: PColor) -> None:
+        print(c.draw(), c.complex_method())
+
+    represent(nice) # OK
+    represent(another) # Also OK
+
+Note that there is no conceptual difference between explicit and implicit
+subtypes, the main benefit of explicit subclassing is to get some protocol
+methods "for free". In addition, type checkers can statically verify that
+the class actually implements the protocol correctly::
+
+    class RGB(Protocol):
+        rgb: Tuple[int, int, int]
+
+        @abstractmethod
+        def intensity(self) -> int:
+            return 0
+
+    class Point(RGB):
+        def __init__(self, red: int, green: int, blue: str) -> None:
+            self.rgb = red, green, blue  # Error, 'blue' must be 'int'
+
+        # Type checker might warn that 'intensity' is not defined
+
+A class can explicitly inherit from multiple protocols and also form normal
+classes. In this case methods are resolved using normal MRO and a type checker
+verifies that all subtyping are correct. The semantics of ``@abstractmethod``
+is not changed, all of them must be implemented by an explicit subclass
+before it could be instantiated.
+
+
+Merging and extending protocols
+-------------------------------
+
+The general philosophy is that protocols are mostly like regular ABCs,
+but a static type checker will handle them specially. Subclassing a protocol
+class would not turn the subclass into a protocol unless it also has
+``typing.Protocol`` as an explicit base class. Without this base, the class
+is "downgraded" to a regular ABC that cannot be used with structural
+subtyping.
+
+A subprotocol can be defined by having *both* one or more protocols as
+immediate base classes and also having ``typing.Protocol`` as an immediate
+base class::
+
+  from typing import Sized, Protocol
+
+  class SizedAndCloseable(Sized, Protocol):
+      def close(self) -> None:
+          ...
+
+Now the protocol ``SizedAndCloseable`` is a protocol with two methods,
+``__len__`` and ``close``. If one omits ``Protocol`` in the base class list,
+this would be a regular (non-protocol) class that must implement ``Sized``.
```

Comment:
It may be clear for a human (taking into account the name of subclass). But for a type checker there are two equally possible options: a user actually wanted a protocol, or just forgot to implement a protocol method. As you said above:
> we don't want to accidentally have some class act as a protocol just because one of its base classes happens to be one

I think explicit is better than implicit applies here.